host/sgx/create.c

```diff
@@ -562,11 +572,117 @@ oe_result_t oe_sgx_validate_enclave_properties(
+
+    if (eeid && eeid->data_size > 0)
+    {
+        sgx_sigstruct_t* sigstruct = (sgx_sigstruct_t*)properties->sigstruct;
+
+        OE_SHA256 ext_mrenclave;
+        oe_sha256_final(&context->hash_context, &ext_mrenclave);
+
+        OE_CHECK(oe_sgx_sign_enclave(
+            &ext_mrenclave,
+            properties->config.attributes,
+            properties->config.product_id,
+            properties->config.security_version,
+            OE_DEBUG_SIGN_KEY,
+            OE_DEBUG_SIGN_KEY_SIZE,
+            sigstruct));
+    }
+
+done:
+    return result;
+}
+
+static oe_result_t _add_eeid_pages(
+    oe_sgx_load_context_t* context,
+    oe_sgx_enclave_properties_t* props,
+    oe_eeid_t* eeid,
+    uint64_t enclave_addr,
+    uint64_t* vaddr,
+    uint64_t entry_point)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    if (eeid)
+    {
+        if (props->header.size_settings.num_heap_pages != 0 ||
+            props->header.size_settings.num_stack_pages != 0 ||
+            props->header.size_settings.num_tcs != 1)
+            OE_RAISE(OE_INVALID_PARAMETER); // For now.
+
+        oe_sha256_context_t* hctx = &context->hash_context;
+        oe_sha256_save(hctx, eeid->hash_state.H, eeid->hash_state.N);
+        eeid->vaddr = *vaddr;
+        eeid->entry_point = entry_point;
+        eeid->signature_size = sizeof(sgx_sigstruct_t); // SGX only for now
+
+        size_t num_bytes = 8 + sizeof(oe_eeid_t) + eeid->data_size;
+        size_t num_pages =
+            num_bytes / OE_PAGE_SIZE + ((num_bytes % OE_PAGE_SIZE) ? 1 : 0);
+
+        sgx_sigstruct_t* sigstruct = (sgx_sigstruct_t*)&props->sigstruct;
+        memcpy(eeid->signature, (uint8_t*)sigstruct, sizeof(sgx_sigstruct_t));
+
+        oe_page_t* pages =
+            oe_memalign(OE_PAGE_SIZE, sizeof(oe_page_t) * num_pages);
+        *((uint64_t*)pages->data) = OE_EEID_MAGIC;
+        memcpy(
+            pages->data + sizeof(uint64_t),
+            eeid,
+            sizeof(oe_eeid_t) + eeid->data_size);
+
+        for (size_t i = 0; i < num_pages; i++)
+        {
+            uint64_t addr = enclave_addr + *vaddr;
+            uint64_t src = (uint64_t)(pages + i * OE_PAGE_SIZE);
+            uint64_t flags = SGX_SECINFO_REG | SGX_SECINFO_R | SGX_SECINFO_W;
+
+            OE_CHECK(oe_sgx_load_enclave_data(
+                context, enclave_addr, addr, src, flags, false));
+            *vaddr += OE_PAGE_SIZE;
+        }
+
+        free(pages);
+
+        // To support non-zero heap/stack images we would have to save the old
+        // sizes here.
+
+        props->header.size_settings.num_heap_pages =
+            eeid->size_settings.num_heap_pages;
+        props->header.size_settings.num_stack_pages =
+            eeid->size_settings.num_stack_pages;
+        props->header.size_settings.num_tcs = eeid->size_settings.num_tcs;
+    }
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+#endif
+
+#ifdef OE_WITH_EXPERIMENTAL_EEID
+oe_result_t oe_sgx_build_enclave(
```

Comment:
I thought this initial design was to make a new API. The API shouldn't change when this flag is enabled

Reply:
I didn't consider this to be part of "the API", but if you prefer, I can turn it into a new function instead of adapting the existing one.

Reply:
Ah i see. Maybe it makes more sense as part of the _oe_sgx_load_context struct, not sure. @CodeMonkeyLeet any thoughts?

Reply:
Yeah, I like that, it would make sense to put it next to the hash context and it avoids changing the function signature.

Reply:
Done. I like this much better too. We could further pass `settings`/`settingscount` to `oe_sgx_initialize_load_context` to initialize `context->eeid` there, but that would touch the header file again.